src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -888,70 +1003,186 @@ scaleCreditsForMerge SingleRun{} _ = ScaledCredits 0
+    -- that is okay because we will find out if\/when we run takeCredits.
+    let unspentCredits' = unspentCredits + c
+
+    -- Try to do some merging work if the new total now goes over the threshold.
+    --
+    -- TODO: when a merge is already completed, then could we skip most of the
+    -- stuff below?
+    when (unspentCredits' >= creditsThresh) $ do
+      isMergeDone <-
+        -- Try to take some unspent credits. The number of taken credits is the
+        -- number of merging steps we will try to do.
+        --
+        -- If an error happens during the body, then we put back as many credits
+        -- as we took, even if the merge has progressed. See Note [Credits] why
+        -- this is okay.
+        bracketOnError (tryTakeUnspentCredits unspentCreditsVar thresh (Credit unspentCredits'))
+                       (mapM_ (putBackUnspentCredits unspentCreditsVar)) $ \case
+          Nothing -> pure False
+          Just c' -> stepMerge mergeVar c'
+
+      -- If we just finished the merge, then we convert the output of the merge
+      -- into a new run. i.e., we complete the merge.
+      --
+      -- If an async exception happens before we get to perform the completion,
+      -- then that is fine. Either the next supplyMergeCredits or
+      -- expectCompletedMerge will complete the merge.
+      when isMergeDone $ completeMerge mergeVar
+
+{-# SPECIALISE tryTakeUnspentCredits ::
+     PrimVar RealWorld Int
+  -> CreditThreshold
+  -> Credit
+  -> IO (Maybe Credit) #-}
+-- | In a CAS-loop, subtract credits from the unspent credits to get it below
+-- the threshold again. If succesful, return Just that many credits, or Nothing
+-- otherwise.
+--
+-- The number of taken credits is a multiple of creditsThresh, so that the
+-- amount of merging work that we do each time is relatively uniform.
+--
+-- Nothing can be returned if the variable has already gone below the threshold,
+-- which may happen if another thread is concurrently doing the same loop on
+-- unspentCreditsVar.
+tryTakeUnspentCredits ::
+     PrimMonad m
+  => PrimVar (PrimState m) Int
+  -> CreditThreshold
+  -> Credit
+  -> m (Maybe Credit)
+tryTakeUnspentCredits !unspentCreditsVar thresh@(CreditThreshold !creditsThresh) (Credit !prev)
+  | prev < creditsThresh = pure Nothing
+  | otherwise = do
+      -- numThresholds is guaranteed to be >= 1
+      let !numThresholds = prev `div` creditsThresh
+      prev' <- casInt unspentCreditsVar prev (prev - numThresholds * creditsThresh)
+      if prev' == prev then
+        pure (Just (Credit prev))
+      else
+        tryTakeUnspentCredits unspentCreditsVar thresh (Credit prev')
+
+{-# SPECIALISE putBackUnspentCredits :: PrimVar RealWorld Int -> Credit -> IO () #-}
+putBackUnspentCredits ::
+     PrimMonad m
+  => PrimVar (PrimState m) Int
+  -> Credit
+  -> m ()
+putBackUnspentCredits !unspentCreditsVar (Credit !x) = void $ fetchAddInt unspentCreditsVar x
```

Comment:
So `fetchAddInt` always succeeds and can't spuriously fail like CAS? Seems like it from the documentation I see, but not totally sure.

Reply:
What do you mean by fail spuriously?

Reply:
I don't know all of the details, but CAS instructions can sometimes fail even if the current value matches the expected one (at least on non-x86). There are often different versions exposed (the strong one compiling to a loop on systems where spurious failures can happen), see e.g. https://en.cppreference.com/w/cpp/atomic/atomic_compare_exchange. But I haven't found documentation on the Haskell implementation. With its API it seems like one wouldn't be able to detect a spurious failure, so I hope it uses the strong version.

Either way, nothing similar is mentioned in https://en.cppreference.com/w/cpp/atomic/atomic/fetch_add, so I think `fetchAddInt` does not have to consider spurious failure.

Reply:
I followed the definitions up to the implementation of the primitive operation underlying `casInt`:

 `casInt -> casIntArray# -> CasByteArrayOp_Int -> __atomic_compare_exchange_n`
 
 And this is the code which seems to implementing the primitive:
 
 https://gitlab.haskell.org/ghc/ghc/-/blob/master/libraries/ghc-prim/cbits/atomic.c?ref_type=heads#L276
 
 `__atomic_compare_exchange_n` is called with `weak=false`, so it is using the strong version. Interesting rabbit hole!